clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchBufferTest.java

```diff
@@ -0,0 +1,200 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * This tests the {@link FetchBuffer} functionality in addition to what {@link FetcherTest} tests during the course
+ * of its tests. One of the main concerns of these tests are that we correctly handle both places that data is held
+ * internally:
+ *
+ * <ol>
+ *     <li>A special "next in line" buffer</li>
+ *     <li>The remainder of the buffers in a queue</li>
+ * </ol>
+ */
+public class FetchBufferTest {
+
+    private final Time time = new MockTime(0, 0, 0);
+    private final TopicPartition topicAPartition0 = new TopicPartition("topic-a", 0);
+    private final TopicPartition topicAPartition1 = new TopicPartition("topic-a", 1);
+    private final TopicPartition topicAPartition2 = new TopicPartition("topic-a", 2);
+    private final Set<TopicPartition> allPartitions = partitions(topicAPartition0, topicAPartition1, topicAPartition2);
+    private LogContext logContext;
+
+    private SubscriptionState subscriptions;
+
+    private FetchConfig<String, String> fetchConfig;
+
+    private FetchMetricsManager metricsManager;
+
+    @BeforeEach
+    public void setup() {
+        logContext = new LogContext();
+
+        Properties p = new Properties();
+        p.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092");
+        p.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
+        p.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
+        ConsumerConfig config = new ConsumerConfig(p);
+
+        Deserializers<String, String> deserializers = new Deserializers<>(new StringDeserializer(), new StringDeserializer());
+
+        subscriptions = createSubscriptionState(config, logContext);
+        fetchConfig = createFetchConfig(config, deserializers);
+
+        Metrics metrics = createMetrics(config, time);
+        metricsManager = createFetchMetricsManager(metrics);
+    }
+
+    /**
+     * Verifies the basics: we can add buffered data to the queue, peek to view them, and poll to remove them.
+     */
+    @Test
+    public void testBasicPeekAndPoll() {
+        try (FetchBuffer<String, String> fetchBuffer = new FetchBuffer<>(logContext)) {
+            CompletedFetch<String, String> completedFetch = completedFetch(topicAPartition0);
+            assertTrue(fetchBuffer.isEmpty());
+            fetchBuffer.add(completedFetch);
+            assertTrue(fetchBuffer.hasCompletedFetches(p -> true));
+            assertFalse(fetchBuffer.isEmpty());
+            assertNotNull(fetchBuffer.peek());
+            assertEquals(completedFetch, fetchBuffer.peek());   // Reference equality
```

Comment:
If we do want to check that the buffer is returning the same reference we should probably use assertSame. AssertEquals could be returning true for 2 different references if we ever override the equals on the CompletedFetch, so the test wouldn't be doing what the comment hints it does. (I would add the assertSame along with the existing assertEquals so we are covered for reference and value)

Reply:
Changed `assertEquals` to `assertSame` as suggested to more clearly convey reference comparison.